File: app/src/main/java/com/example/java_viewmodel/mainlist/ListViewModel.java
package com.example.java_viewmodel.mainlist;

import android.app.Application;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.example.java_viewmodel.contents.ItemDatabaseRepository;

import java.util.List;

public class ListViewModel extends AndroidViewModel {

    private final ItemDatabaseRepository itemDatabaseRepository;

    private List<ListItemValue> itemList;
    private final MutableLiveData<List<ListItemValue>> listMutableLiveData = new MutableLiveData<>();

    public ListViewModel(@NonNull Application application) {
        super(application);
        itemDatabaseRepository = new ItemDatabaseRepository(application);
        LiveData<List<ListItemValue>> initItemValues = itemDatabaseRepository.getAllItemValues();
        initItemValues.observeForever(initValueList -> {
            itemList = initValueList;
            listMutableLiveData.postValue(initValueList);
        });
    }

    public List<ListItemValue> getItemList() {
        return itemList;
    }

    public void setItemList(List<ListItemValue> itemList) {
        this.itemList = itemList;
        listMutableLiveData.setValue(itemList);
    }

    public boolean isEmpty() {
        return itemList == null || itemList.isEmpty();
    }

    public LiveData<List<ListItemValue>> getLiveData() {
        return listMutableLiveData;
    }

    public int getItemIndex(@NonNull String number) {
        if (itemList == null) {
            return -1;
        }
        for (int i = 0; i < itemList.size(); i++) {
            if (number.equals(itemList.get(i).getNumber())) {
                return i;
            }
        }
        return  -1;
    }

    public boolean decreaseItemCount(int index, int count) {
        if (itemList == null) {
            return false;
        }
        ListItemValue item = itemList.get(index);
        if (item.getCurrentCount() < count) {
            return false;
        }
        item.setCurrentCount(item.getCurrentCount() - count);
        return true;
    }

    public void populateData() {
        itemDatabaseRepository.populateItemData();
    }

    public void saveList() {
        Log.d("ListViewModel", "numbers=" + String.join(",", listMutableLiveData.getValue().stream().map(o -> String.valueOf(o.getCurrentCount())).toArray(String[]::new)));
        itemDatabaseRepository.clearInsertList(listMutableLiveData.getValue());
    }

//    public static class ListViewModelFactory implements ViewModelProvider.Factory {
//
//        @NonNull
//        private final Application application;
//
//        private final long minInitCount;
//
//        public ListViewModelFactory(@NonNull Application application, long minInitCount) {
//            this.application = application;
//            this.minInitCount = minInitCount;
//        }
//
//        @SuppressWarnings("unchecked")
//        @NonNull
//        @Override
//        public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
//            if (modelClass == ListViewModel.class) {
//                return (T) new ListViewModel(application, minInitCount);
//            } else {
//                return null;
//            }
//        }
//    }
}

File: app/src/main/java/com/example/kt_viewmodel/mainlist/ListViewModel.kt
package com.example.kt_viewmodel.mainlist

import android.util.Log
import androidx.lifecycle.LiveData
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import com.example.kt_viewmodel.contents.ItemDatabaseRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import javax.inject.Inject

@HiltViewModel
class ListViewModel @Inject constructor(
    private val itemDatabaseRepository: ItemDatabaseRepository,
    private val savedStateHandle: SavedStateHandle
) : ViewModel() {
    private val _itemList = itemDatabaseRepository.getAllItemValues()
    val itemList: LiveData<List<ListItemValue>>
        get() = _itemList

    fun getItemIndex(number: String): Int {
        val list = _itemList.value ?: return -1
        for (i in list.indices) {
            if (number == list[i].number) {
                return i
            }
        }
        return -1
    }

    fun decreaseItemCount(index: Int, count: Int): Boolean {
        val list = _itemList.value ?: return false
        val item = list[index]
        if (item.currentCount < count) {
            return false
        }
        item.currentCount = item.currentCount - count
        return true
    }

    fun populateData() {
        itemDatabaseRepository.populateItemData()
    }

    fun saveList() {
        Log.d("ListViewModel", "numbers=" + _itemList.value?.joinToString(",") { o -> o.currentCount.toString() })
        itemDatabaseRepository.clearInsertList(_itemList.value)
    }
}